### PR TITLE
Add LeakyReLu implementation

### PR DESCRIPTION
**Pull request by KarstenB**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed. All complained about a typo and a non-working unit test (that is already failing on main for me). But std and no-std both worked.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

I implemented this because I received the error message in #1130. I then figured out that for my Model LeakyReLu was missing.

### Changes

I implemented the LeakyReLu activation function. I have no idea if the way it is done is the most efficient, but it gets the job done. 

### Testing

I wrote a small unit test in my code for the new LeakyRelu to ensure that it works as intended. I couldn't find a good place within the burn framework, so I would be happy to put it there if someone points me to it.

Also my model that is using the ReLu produces the exact same values as the python version.

### CODE REVIEW

**Review by antimora — COMMENTED** [review]
Thanks for the addition. Could you also add the following tests?

1. ONNX file testing. Please see this crate: https://github.com/tracel-ai/burn/tree/main/burn-import/onnx-tests
2. Tensor test: https://github.com/tracel-ai/burn/tree/main/burn-tensor/src/tests/activation (use this PR on how to add test: https://github.com/tracel-ai/burn/pull/260/files). To test tensor crate, you'll to cd one of the backends, e.g. burn-ndarray and run test (cargo test). It will import the tensor tests.

**Comment by antimora** [comment]
Regarding the typo test fail. We fixed it recently, please update from the main.

**Comment by antimora** [comment]
@nathanielsimard do we add a module for LeakyRelu as well just like other activation functions. I was looking at our book on tensor and I couldn't see anything on activation functions for the tensor but only for nn (https://burn.dev/book/building-blocks/module.html#built-in-modules). 

If so, we also need to update our book here: https://github.com/tracel-ai/burn/blob/main/burn-book/src/building-blocks/module.md#built-in-modules

**Comment by antimora** [comment]
Code coverage can guide what should be tested.

File: burn-tensor/src/tensor/api/float.rs
Comment (nathanielsimard) [inline burn-tensor/src/tensor/api/float.rs:262]: Activation functions aren't callable on float tensors, so I would remove that function.

File: burn-tensor/src/tensor/activation/base.rs
Comment (nathanielsimard) [inline burn-tensor/src/tensor/activation/base.rs:15]: The backend function should be called instead:

```rust
B::leaky_relu(tensor, alpha)
```

File: burn-tensor/src/tensor/ops/activation.rs
Comment (nathanielsimard) [inline burn-tensor/src/tensor/ops/activation.rs:30]: This seems wrong, the shapes of `ones` and `alpha` should be the same as the inputs.

```rust
        let mask = B::float_lower_elem(tensor.clone(), 0.elem());
        let alpha_tensor = B::float_mul(tensor.clone(), alpha.elem());
        
        // Update the tensor where the values are `> 0` by `tensor * alpha`.
        B::float_mask_where(tensor, mask, alpha_tensor)
```

**Review by nathanielsimard — CHANGES_REQUESTED** [review]

File: burn-tensor/src/tensor/activation/base.rs
Comment (KarstenB) [reply burn-tensor/src/tensor/activation/base.rs]: Done, except for the fact that I had to do some more conversions now.

File: burn-tensor/src/tensor/ops/activation.rs
Comment (KarstenB) [reply burn-tensor/src/tensor/ops/activation.rs]: Done, but used float_mul_scalar

**Comment by KarstenB** [comment]
Thanks for the feedback. I applied it as good as I can, but I am not sure I understand the comment about the Module. Let me know if there is anything more that I need to do.

File: burn-tensor/src/tensor/api/float.rs
Comment (KarstenB) [reply burn-tensor/src/tensor/api/float.rs]: Done

File: burn-tensor/src/tensor/ops/activation.rs
Comment (KarstenB) [inline burn-tensor/src/tensor/ops/activation.rs:48]: I am not quite sure whether I should take grad here instead of output...

Comment (KarstenB) [reply burn-tensor/src/tensor/ops/activation.rs]: That most definitely should be grad.

File: burn-tensor/src/tensor/ops/activation.rs
Comment (nathanielsimard) [inline burn-tensor/src/tensor/ops/activation.rs:47]: I feel like this is maybe wrong. This is supposed to be the same mask as during the forward pass, but it's calculated on the output? Maybe I'm wrong though. @louisfd 

I think this will work if the alpha is a positive float, but otherwise we lose the sign information.

Comment (antimora) [reply burn-tensor/src/tensor/ops/activation.rs:47]: tagging @louisfd

Comment (louisfd) [reply burn-tensor/src/tensor/ops/activation.rs:47]: You're right Nathaniel, if alpha<0 this won't work. However I don't think LeakyRelu is ever used with a negative alpha. We _could_ maybe assert that alpha is positive. 
But, since it would be easy to save the input as state instead of the output to cover the general case, I say we do that.

**Review by louisfd — CHANGES_REQUESTED** [review]

File: burn-tensor/src/tensor/ops/activation.rs
Comment (KarstenB) [reply burn-tensor/src/tensor/ops/activation.rs:47]: I know that other libraries are checking for a negative Alpha and raising an exception if they encounter one. The reasoning for that is that the autodiff could not handle then negative case. I am fine with implementing either one. 

Just to be sure, you mean to implement it like this:
```rust
let mask = B::float_lower_elem(grad.clone(), 0.elem());
```
correct?

Comment (louisfd) [reply burn-tensor/src/tensor/ops/activation.rs:47]: Hi @KarstenB 
Not exactly, grad doesn't carry that information. For this function it's only a matter of renaming the argument `output` to `input`. Then it's where we call `leaky_relu_backward` that we must pass the input instead of the output. 

I think you haven't done that part yet but it should be easy since you can copy relu. In `crates/burn-autodiff/src/ops/activation.rs` you'll find the relu function, you can copy it and straightforwardly rename `relu` occurrences to `leaky_relu`. 

Then what matters is the `state` variable. At the bottom you'll find `let state = prep.checkpoint(&tensor);` which means the state is actually the input. Above you'll find `B::relu_backward(state, grad)`, so `state` is the input. I realize there's actually a misnaming for relu as well, because the input is given but `relu_backward`'s first arg is named output. It matters less for relu because the input is <0 and the output is ==0, and in relu_backward we use lower or equal. Because relu_backward already takes the input then you won't have to change anything else than the names in the autodiff function.

**Comment by antimora** [comment]
LeakyRelu module was added. This RP will be about ONNX OP addition. I can update the outstanding changes and merge them.

**Comment by KarstenB** [comment]
> LeakyRelu module was added. This RP will be about ONNX OP addition. I can update the outstanding changes and merge them.

Feel free to use my code in whatever way you want to! I currently don't have the bandwidth to continue working on this, but will be extremely happy when it exists as a feature :)

**Comment by antimora** [comment]
OK. I have kept the new activation function minus `backward` impl. I updated the prior LeakyRelu module with the new activation function.

@nathanielsimard , @louisfd, please review it again.

**Review by antimora — APPROVED** [review]
I made the necessary fixes and now we need more approval from others.

File: crates/burn-tensor/src/tensor/ops/activation.rs
Comment (laggui) [inline crates/burn-tensor/src/tensor/ops/activation.rs:28]: The comment seems incorrect but the actual implementation looks fine.

I think it should be: "where the values are `< 0`"

**Review by laggui — APPROVED** [review]
Implementation LGTM! Just a minor change request related to a comment but I'm approving in advance.

**Review by laggui — APPROVED** [review]

**Review by nathanielsimard — APPROVED** [review]

**Comment by KarstenB** [comment]
@antimora Thanks a lot for finishing this up! Really grateful to see this being merged!

**Comment by antimora** [comment]
@KarstenB Thank you for your contribution! Ended up using all of your code - just had to align with the existing code.

**Pull request merged**
